src/scenechange/mod.rs

```diff
@@ -7,69 +7,202 @@
+  excluded_frames: BTreeSet<u64>,
 }
 
-impl<T: Pixel> Default for SceneChangeDetector<T> {
-  fn default() -> Self {
+impl SceneChangeDetector {
+  pub fn new(bit_depth: u8) -> Self {
+    // This implementation is based on a Python implementation at
+    // https://pyscenedetect.readthedocs.io/en/latest/reference/detection-methods/.
+    // The Python implementation uses HSV values and a threshold of 30. Comparing the
+    // YUV values was sufficient in most cases, and avoided a more costly YUV->RGB->HSV
+    // conversion, but the deltas needed to be scaled down. The deltas for keyframes
+    // in YUV were about 1/3 to 1/2 of what they were in HSV, but non-keyframes were
+    // very unlikely to have a delta greater than 3 in YUV, whereas they may reach into
+    // the double digits in HSV. Therefore, 12 was chosen as a reasonable default threshold.
+    // This may be adjusted later.
+    const BASE_THRESHOLD: u8 = 12;
     Self {
-      // This implementation is based on a Python implementation at
-      // https://pyscenedetect.readthedocs.io/en/latest/reference/detection-methods/.
-      // The Python implementation uses HSV values and a threshold of 30. Comparing the
-      // YUV values was sufficient in most cases, and avoided a more costly YUV->RGB->HSV
-      // conversion, but the deltas needed to be scaled down. The deltas for keyframes
-      // in YUV were about 1/3 to 1/2 of what they were in HSV, but non-keyframes were
-      // very unlikely to have a delta greater than 3 in YUV, whereas they may reach into
-      // the double digits in HSV. Therefore, 12 was chosen as a reasonable default threshold.
-      // This may be adjusted later.
-      threshold: 12,
-      last_frame: None,
+      threshold: BASE_THRESHOLD * bit_depth / 8,
+      excluded_frames: BTreeSet::new(),
     }
   }
-}
 
-impl<T: Pixel> SceneChangeDetector<T> {
-  pub fn new(bit_depth: usize) -> Self {
-    let mut detector = Self::default();
-    detector.threshold = detector.threshold * bit_depth as u8 / 8;
-    detector
+  /// Runs keyframe detection on the next frame in the lookahead queue.
+  ///
+  /// This function requires that a subset of input frames
+  /// is passed to it in order, and that `keyframes` is only
+  /// updated from this method. `input_frameno` should correspond
+  /// to the first frame in `frame_set`.
+  ///
+  /// This will gracefully handle the first frame in the video as well.
+  pub fn analyze_next_frame<T: Pixel>(
+    &mut self,
+    previous_frame: Option<Arc<Frame<T>>>,
+    frame_set: &[Arc<Frame<T>>],
+    mut input_frameno: u64,
+    config: &EncoderConfig,
+    inter_cfg: &InterConfig,
+    keyframes: &mut BTreeSet<u64>,
+  ) {
+    let frame_set = match previous_frame {
+      Some(frame) => [frame].iter().chain(frame_set.iter()).cloned().collect(),
+      None => {
+        // The first frame is always a keyframe.
+        keyframes.insert(0);
+        input_frameno += 1;
```

Comment:
Shouldn't this just return? It seems to me that currently the second frame is analyzed twice (first time through here and second time normally).

Reply:
The reason this doesn't return here is because it still needs to call `self.is_key_frame` to see if any of the frames after it can be excluded as flashes. But I'm realizing the names of these functions make it really unclear as to what is going on. I'll rename them to see if that makes things clearer.

Reply:
Actually, maybe you are right that it should just return...